Signal/src/ViewControllers/ContactViewController.swift

```diff
@@ -0,0 +1,686 @@
+        super.init(frame: CGRect.zero)
+
+        self.isUserInteractionEnabled = true
+        self.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(wasTapped)))
+    }
+
+    func wasTapped(sender: UIGestureRecognizer) {
+        Logger.info("\(self.logTag) \(#function)")
+
+        guard sender.state == .recognized else {
+            return
+        }
+        actionBlock()
+    }
+}
+
+// MARK: -
+
+class ContactViewController: OWSViewController, CNContactViewControllerDelegate {
+
+    let TAG = "[ContactView]"
+
+    enum ContactViewMode {
+        case systemContactWithSignal,
+        systemContactWithoutSignal,
+        nonSystemContact,
+        noPhoneNumber,
+        unknown
+    }
+
+    private var hasLoadedView = false
+
+    private var viewMode = ContactViewMode.unknown {
+        didSet {
+            SwiftAssertIsOnMainThread(#function)
+
+            if oldValue != viewMode && hasLoadedView {
+                updateContent()
+            }
+        }
+    }
+
+    let contactsManager: OWSContactsManager
+
+    var reachability: Reachability?
+
+    override public var canBecomeFirstResponder: Bool {
+        return true
+    }
+
+    private let contact: OWSContact
+
+    // MARK: - Initializers
+
+    @available(*, unavailable, message: "use init(call:) constructor instead.")
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("Unimplemented")
+    }
+
+    required init(contact: OWSContact) {
+        contactsManager = Environment.current().contactsManager
+        self.contact = contact
+        self.scrollView = UIScrollView()
+
+        super.init(nibName: nil, bundle: nil)
+
+        tryToDetermineMode()
+
+        NotificationCenter.default.addObserver(forName: .OWSContactsManagerSignalAccountsDidChange, object: nil, queue: nil) { [weak self] _ in
+            guard let strongSelf = self else { return }
+            strongSelf.tryToDetermineMode()
+        }
+
+        reachability = Reachability.forInternetConnection()
+
+        NotificationCenter.default.addObserver(forName: .reachabilityChanged, object: nil, queue: nil) { [weak self] _ in
+            guard let strongSelf = self else { return }
+            strongSelf.tryToDetermineMode()
+        }
+    }
+
+    // MARK: - View Lifecycle
+
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        UIUtil.applySignalAppearence()
+
+        self.becomeFirstResponder()
+
+        contactsManager.requestSystemContactsOnce(completion: { [weak self] _ in
+            guard let strongSelf = self else { return }
+            strongSelf.tryToDetermineMode()
+        })
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+
+        UIUtil.applySignalAppearence()
+
+        self.becomeFirstResponder()
```

Comment:
Why are we becoming first responder here?

If it's because of https://github.com/signalapp/Signal-iOS/pull/3333, then I'd prefer we figure out whats going on rather than continuing to spread around this partial work-around, in case it has unintended consequences.

Reply:
Removed.